### ISSUE TITLE
Support working with huge parameter spaces

### ISSUE THREAD

**Issue by JohanMollevik** [autonomio/talos#201]
I am experimenting with setting up talos to explore the hyperparameter space of my problem.
Naively I thought that I could just add mor or less continous ranges of my hyperparameters and let talos random sampling and round_limit sort out the huge parameter space, 10^16 premutations by my estimate.

Doing this I wrote this parameter list

        params = {
                'epoch': [2],
                'batch_size': range(1,128,1),
                'activation': [relu], # TODO use for all layers or define different
                # convultion layers in the begining
                'conv_hidden_layers': range(1,4,1),
                'conv_depth_shape': ['funnel','long_funnel','rhombus','brick','diamond','hexagon','triangle','stairs'],
                'conv_size_shape': ['funnel','long_funnel','rhombus','brick','diamond','hexagon','triangle','stairs'],
                'conv_depth_first_neuron': range(10,100,1),
                'conv_depth_last_neuron': range(5,100,1),
                'conv_size_first_neuron': range(3,15,1),
                'conv_size_last_neuron': range(3,15,1),
                # fully connected layers at the end
                'first_neuron': range(2,128,1),
                'last_neuron': range(2,128,1),
                'shapes': ['funnel','long_funnel','rhombus','brick','diamond','hexagon','triangle','stairs'],
                'hidden_layers': range(0,5,1),
        }

which did not work that well. Looking at the talos source it seems that this line in in talos/parameters/ParamGrid.py is where it fails

    _param_grid_out = array(list(product(*ls)), dtype='object')

When this tries to build the premutations explicitly in the list function it runs out of memmory and fails.

Can we have a feature where we do not build up the parameter space in memmory?

(I will use some workaround for now so consider this a feature request)

**Comment by mikkokotila** [comment]
Note the section in the docs related with [parameters dictionary](https://autonomio.github.io/docs_talos/#parameter-dictionary). Also, to avoid doubt, you have three meaningful ways to input parameters: 

- as a list with more than one value e.g [1, 2, 3] which will add values 1, 2, 3
- as a tuple with start, end, no_of_steps e.g. (0, 1, 6) which will add values 0.0, 0.1, 0.2, 0.3, 0.4, 0.5
- as a list with a single value e.g. [200] which will add 200

That said, when you say "build parameter space in memory", what do you suggest instead of memory?

**Comment by JohanMollevik** [comment]
I think I got the use of parameter dictioneries right.

What I mean by not in memmory is instead of building the param_grid list talos could save the ls list and when a new configuration is needed randomly select each dimension separatly (saving the used dimensions in a used list and trying again if the same one is selected).

This obviously need some care when working with different random functions

Some pseudocude

    params=ls # the ls from the code snippet I quoted
    while selected_parameters not in blocklist:
        selected_parameters=[random.choice(parameter_options) for parameter_options in params]
        blocklist.add(selected_parameters)
    return selected_parameters

The idea is the same as that of pythons built in range function, lazily represent the parameterspace and only read out the parts we care about.

**Comment by mikkokotila** [comment]
I'm open to this. I think it's attractive to be able to handle very large parameter spaces. That said, there are certain benefits from the ability of having an index number of some sort upon which a given permutation is returned. That allows agnostically using any random function to make the picks. Do you have idea how we could retain this in something like you propose?

**Comment by JohanMollevik** [comment]
A simple way is to have the inedx number be a tuple/list/string of integers with a length the same as the number of parameters.

Thus

    params =  {
        'batch_size': [1,16,32,128],
        'first_neuron': [4,16,64],
        'hidden_layers': [0,1,3],
        'dropout': [0,0.2]
    }

Could have an index (16,4,1,0) or [16,4,1,0] or '16_4_1_0' for producing the premutation with
batch_size: 16
first_neuron: 4
hidden_layers: 1
droppout: 0

this can ofcourse be mapped into an integer bu I think leaving it in tupple/list/string form makes it clearer

The disadvantage is of cource that the random function have to produce a tuple, this can be handled either by calling the random function several times, one for each index (will that mess with distribution?) or by having a mapping function from integers to such a tupple (again, distributions).

We can create such a mapping for example like this.

Given a list of the length of each parameter and the integer to map

loop over the list from end
* compute the number of digits of that parameters length as n
* save the n last digits of the integer as the index for the last parameter
* remove the n last digits from the integer
* repeat for then next parameter

Doing any of this will cause problems for a normal distributed random function and similar, but on the other hand I do not know if that is well behaved with the current code either.

**Comment by mikkokotila** [comment]
Right now the way things work is that we have to important components to consider: 

1) all possible permutations 
2) the simplest possible index 

The index serves the purpose of keeping track of permutations that have been already explored, or for some other reason are to be ignored. This approach is beneficial in two important ways. 

a) any function, random or otherwise, that outputs an integer can be used to pick or ignore items in the permutations 
b) we entirely avoid the need for performing operations in the permutation space (which might be big) 

Out of all components / decisions in Talos, this is the one that is most connected with the mainline program. The current approach, related with (b), had been selected with great care to allow very large permutation space. It's meaningful to know that it fails at 10^16. So the first action item here would be to find out at which order of magnitude things stop working. 

This is just to share with you the background i.e. whatever is in place, had been selected for a reason, and any changes to the parameter grid aspect are closely connected with the mainline program (as well some important ideas about future development). 

I'm not averse to well structured proposals for something better, as long as those proposals appreciate the background I had provided. Obviously there are finer details that would have to be discussed as part of the process. If you are interested in contributing, I whole heartedly welcome it. 

Related with your specific case in the example, I think that with number of neurons a strict one-by-one progression in parameters might not be warranted. In your case, just by changing that your case falls within the supported order of magnitude.

**Comment by mikkokotila** [comment]
To update, I did some quick test and it does appear that around 10^8 things become slow in a low end laptop machine. But just for the period `_param_grid_out = array(list(product(*ls)), dtype='object')` is run as immediately upon completion the memory is released mostly. It took about 1 minute to process with a machine that has few gigs of available ram and low-end CPU. 

I'm thinking if there is some meaningful way we can support large permutation spaces without changing anything else i.e. is there a step of random picking we can add before we run `_param_grid_out = array(list(product(*ls)), dtype='object')` that reduces the workload down to say 10^8. Because we are anyway downsampling, it seems to make no difference if we can come up with a meaningful way to do that. 

What do you think? 

I'm going to label this as "must have" for the part that we should support large theoretical parameter grids. Even much larger than 10^16.

**Comment by JohanMollevik** [comment]
Yes that matches my observations, I think that doing a partial down sampling before the product is an idea worth exploring, if it works great otherwise some other approch is needed.

**Comment by mikkokotila** [comment]
I think we will be able to resolve this in a meaningful way by having such a preliminary reducer for very large permutation spaces. The flow control is tricky though, so will take some time to program. For example, we might have a parameter with 10000 values and another one with 3 values, so there needs to be a way that does not reduce the one with 3 at all, but randomly picks the maximum number of values from the one with 10000.

**Comment by JohanMollevik** [comment]
I agree about the tricky part, what we have to be carefull about is the case with many parameters but not that many options each as doing the reduction before the product will force the samples onto a virtual grid which defeats the stregths of random a bit. If I get the time I might try to write a replacement for the product function that internalizes the randomization (using a passed in randomizer taking integers), if I get that working we can see if we can get that to fit into the overall architechture or if another approach is needed.

**Comment by JohanMollevik** [comment]
I have some time today so I will take a stab at this, the idea I'm persuing hinges on the assumption I make in #200 that it is ok to move the handling of round_limit to ParamGrid, if there are issues with that please let me know so that I do not waste work.

**Comment by mikkokotila** [comment]
Some background that might be helpful. 

1) it's important that we have a single integer index, one value per permutation
2) downsampling has to support a variety of random methods
3) #200 I believe is a very easy and low risk fix (as opposed to this being the single highest risk item)

I can provide more details on each of course, but those would be the headline items that are related with this.

**Comment by JohanMollevik** [comment]
Sounds good

If my idea works the change will be rather small and not modify the logic at all, simply reorder and move some parts around. I will see if that pans out thou. 

I will try to get it implemented so I can show you, that is probably the simplest way to explain.

**Comment by JohanMollevik** [comment]
I got something mostly working
I have run into one issue tou

the method

chances.Randomizer

ignores the length parameter

right now it returns a random premotation of the indices up to n

I would need a list of length x with non duplicate values between 0 and n

so for the call 

chances.Randomizer(3,150)

instead of
[0 2 1 ... ]
or
[0 2 1]

it could return

[105 65 4]

What would be the best way to fix that, should it be fixed in the chances library or in talos?

**Comment by mikkokotila** [comment]
This will be taken care of on the chances level. I can do it in approximately 1 hour from now. Should be a quick fix.

**Comment by mikkokotila** [comment]
I checked it and I believe what you are looking for is: 

```
r = chances.Randomizer(150, 3)
r.uniform_mersenne()
```
This will yield something like:

`[138, 116, 27]`

That said, I'm going to add a docstring to chances.Randomizer() that makes it clear.

**Comment by JohanMollevik** [comment]
Which commit version of chances are you looking at? When I look at the code on github the only use of the length parameter is in the statement

self.length=length

**Comment by mikkokotila** [comment]
I  have tested both production (pypi) and current master. 

Note the difference here `chances.Randomizer(3, 150)` is not giving you what you want, but `chances.Randomizer(150, 3)` will give you what you want.

**Comment by JohanMollevik** [comment]
I have something working now

This is enough to demonstrate the concept and find any issues with it but not clean enough to apply yet.

I also noticed a performance bug regarding chances but lets take that as a separate issue (https://github.com/autonomio/chances/issues/2).

[preliminary_patch.txt](https://github.com/autonomio/talos/files/2884692/preliminary_patch.txt)

Two questions
1) What is the if statment with self.main_self.shuffle is True suposed to do, won't it be random anyways?
2) what is the if random_method == 'stratified': block doing, the return format is not the same as for the rest of the code, is this even used anymore?

Other than that, how does it look? Is the aproach workable?

**Comment by mikkokotila** [comment]
Amazing. I will have a look.

Can you: 

1) fork Talos
2) implement the changes to daily-dev branch
3 make pr from your daily-dev to talos/daily-dev

It will be straightforward to review like that.

**Comment by JohanMollevik** [comment]
Ok, pull request created

If you see anything I should explain or look at just let me know

**Comment by JohanMollevik** [comment]
Ok, new pull request, let see if I managed not to mess up this time

**Comment by mikkokotila** [comment]
> 1. What is the if statment with self.main_self.shuffle is True suposed to do, won't it be random anyways?

I have to take a look, but if I remember there are cases where it would not be random already. 

> 2. what is the if random_method == 'stratified': block doing, the return format is not the same as for the rest of the code, is this even used anymore?

Indeed. Tentatively (based on a very quick glance) this has to change:

`return vstack((train, test))`

...to this:

`out = [i[0] for i in vstack((train, test))[:, -1:]]`

**Comment by JohanMollevik** [comment]
Regarding 2, is the intent to only select samples from the training part of sample space? If so, are there any way to access the test part later?

**Comment by mikkokotila** [comment]
This is a really cool upgrade :) Will keep testing (outputs and such) but now available in daily-dev as of v.0.4.9 so closing here. 

Thanks again for a great contribution!

**Comment by JohanMollevik** [comment]
Thanks yourself for providing this package
